proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,553 @@
+backtrace reporting (e.g. the Rust runtime enables debugging with
+``RUST_BACKTRACE=1``). It would be straightforward to provide a utility (either
+in a third-party library or perhaps ``base`` itself) which would configure the
+global backtrace mechanism from the environment: ::
+
+    setBacktraceMechanismFromEnv :: IO ()
+    setBacktraceMechanismFromEnv =
+        getEnv "GHC_BACKTRACE" >>= setEnabledBacktraceMechanisms . parseBacktraceMechanisms
+
+This could be called during program initialization, providing the ease of
+configuration found in other languages. As it could be added at any time,
+``setBacktraceMechanismFromEnv`` is not part of the scope of this proposal.
+
+
+Effects and Interactions
+------------------------
+
+The described mechanism provides users with a convenient means of gaining greater
+insight into the sources of exceptions. Currently the ``+RTS -xc``
+runtime system flag provides an ad-hoc mechanism for reporting exception
+backtraces using the cost-center profiler. While the ``-xc`` mechanism is
+largely subsumed by the mechanism proposed here, we do not propose to remove it
+in the near future.
+
+During discussions on a previous iteration of this proposal, various community
+members mentioned that they were using dynamically-typed annotations on
+exceptions in their own code-bases to great effect. One such library,
+``annotated-exception``, served as the inspiration for the annotation notion
+proposed above and could likely be largely superceded by
+``ExceptionAnnotation``.
+
+Costs and Drawbacks
+-------------------
+
+The introduction of exception context adds a bit of complexity to GHC's
+exception machinery in exchange for a significant improvement in observability.
+
+All-in-all, GHC's exception interface grows considerably under this proposal,
+even if we don't provide every possible variant. Moreover, these changes will
+need to be mirrored in downstream packages (e.g. ``exceptions``).
+
+Moreover, the general nature of exception context slightly muddies the waters
+when it comes to exception hierarchy design. Library authors now have two ways
+of conveying failure information to the caller: they may introduce a new
+exception type (as they can do today) or they can augment an existing exception
+type via the context field. Correctly choosing from between these options may
+be, in some cases, non-obvious and could require an element of design taste.
+
+The introduction of the global state for backtrace mechanism selection is quite
+ad-hoc. We consider this approach to be a compromise which makes robust
+backtraces available by default with minimal additional code. Exception
+backtraces are primarily a debugging tool and are a cross-cutting concern. The
+global backtrace mechanism selection facility proposed here recognizes this but
+it suffers from the usual drawbacks associated with global state: it does not
+compose well and may result in surprising behavior when manipulated by more
+than one actor.
```

Comment:
I think this is a reasonable, pragmatic compromise. Does the proposal mention what the default state for the enabled backtraces is if the user doesn't do anything? I couldn't immediately see this, but I might have missed it.